New `Array#resize` method

beta-ziliani opened the pull request:
Solves #8209 by creating a new `resize` method.

```crystal
ary = ['a', 's', 't', 'e', 'r', 'i', 't', 'e']
p ary.remaining_capacity  # => 8
ary.resize(20)
p ary.remaining_capacity  # => 20
ary.resize(8)
p ary.remaining_capacity  # => 8
```

straight-shoota commented on spec/std/array_spec.cr:
```diff
@@ -2041,4 +2041,18 @@ describe "Array" do
+    ary.resize(3).should eq([2, 3])
+    ary.remaining_capacity.should eq(2)
+
+    expect_raises(ArgumentError, "Insufficient capacity: 2 cannot hold 2 elements with an offset of 1") do
```
This error seems very unexpected. When I have an array of size `2` I would assume it can be resized to capacity `2`. IMO `ary.resize(ary.size)` should always be a valid operation to shrink the capacity to match the current occupancy.

In fact, we could even consider adding an argless overload (or a default value) for exactly this use case. I think it might be a common story: After doing some stuff that dynamically builds the array while adding and maybe removing items from it, at some point all processing is done and the array is not going to change. Then it might make sense to resize capacity to the current size. If there have been a lot of removals during the creation process, that could free a lot of memory.

beta-ziliani replied:
My initial code actually had a default value for this, but I decided against it for it being costly: you need to first move the elements to the beginning of the array, and then perform the resize. I thought of `resize` as a lower-level primitive, only for the expert, who can then decide how to act in such cases. But I guess it makes sense to have it nevertheless.

wontruefree commented on spec/std/array_spec.cr:
```diff
@@ -2041,4 +2041,18 @@ describe "Array" do
     end
   end
+
+  it "#resize" do
```
This includes 2 different test cases.
Could we split this up.  I look through the specs sometimes when something breaks and it is not always obvious what it is trying to test.

oprypin commented:
I don't think `resize` is a good name, because at least in C++, C#, Python+NumPy that means actually dropping elements or filling the array with actual new elements.
C++ has `reserve` (which, by the way, actually can't release memory) and `shrink_to_fit` for this purpose.
And perhaps this pull request's title should mention what the method does.

Fryguy commented:
Perhaps `.capacity=` (maybe with a getter `.capicity`)?

straight-shoota commented:
I agree that `rescale` might be confusing about the semantics. `#capacity=` looks better. But it's a setter, not an action method, so that's a bit of a down side as well. It doesn't immediately convey that it might actually move things around.
Maybe `#rescale` could be an option? Or `#resize_capacity`?

oprypin commented:
FWIW doing the API exactly as in C++ would solve *all* of the issues named so far. It also prevents having to think what happens when you specify a smaller capacity than now. Also I think it's very pragmatic or goal-oriented. I don't often see the need to set a specific *smaller* buffer size.
So, `reserve(size)` (no decrease)

beta-ziliani commented:
If we go with two different methods, I like Java's [`ArrayList`](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html)  `ensureCapacity` and `trimToSize` better. C# also uses `trimToSize`, but for the capacity it has a setter as proposed by @Fryguy (that also shrinks). 

But I don't think the setter is a good idea, since as mentioned in the [comment](https://github.com/crystal-lang/crystal/pull/11485#discussion_r755521471) we need to do something about the offset, requiring an extra (defaultable) arg.

straight-shoota commented on spec/std/array_spec.cr:
```diff
@@ -2042,17 +2042,76 @@ describe "Array" do
+    end
+
+    it "raises if not enough capacity" do
+      expect_raises(ArgumentError, "Insufficient capacity: 2 cannot hold 3 elements") do
+        ary = [1, 2, 3]
+        ary.ensure_capacity(2)
```
Again (https://github.com/crystal-lang/crystal/pull/11485#discussion_r755521471) this seems very unexpected: Why would this method raise? We ask to ensure the array has at least a capacity of 2, and the actual capacity is 3. So everything should be good, nothing to do.

beta-ziliani replied:
Agh, sorry I missed this in the original comment. Will fix.

straight-shoota commented on src/array.cr:
```diff
@@ -2029,6 +2029,36 @@ class Array(T)
+
+  # Reduces the internal buffer to exactly fit the number of elements in the
+  # array. If `rewind` is truthy it moves the elements in the internal array to the front.
+  def trim_to_size(rewind = true) : self
```
I'm not convinced about the usefulness of the `rewind` parameter. `#trim_to_size` should always rewind. I can't see a relevant usecase where you would not want that, especially since you have no external insight on or control of the offset.
For `#ensure_capacity` it's not that clear, but I don't know why it should bother about that either.

carlhoerberg replied:
Agree, without knowing the internals of `Array` no one would know what "rewind" implies

carlhoerberg replied:
would it be useful to have an "extra capacity" argument, that is `resize_to_capacity(size + extra_capacity)`? I think we would want that, so that if an element was added immediately after the trimming, it wouldn't needed to increase capacity so close again. on the other hand i guess `ensure_capacity` can be used for that.

straight-shoota replied:
Yeah, I think the use case for `#trim_to_size` would be to say I'm done with adding or removing things, let's have it trimmed to save space.

beta-ziliani replied:
As it is implemented here, `ensure_capacity` won't shrink, so it won't help you in that case. This is why my original proposal had just one method to set the capacity, either expanding or shrinking it. I wonder what's best, if to go back to the original proposal or to add an extra-capacity argument to leave some room.

beta-ziliani replied:
I wonder if fixing https://github.com/crystal-lang/crystal/issues/11445 would help here in your use case @carlhoerberg .

carlhoerberg replied:
no, it's a very good addition, but our deques/arrays also shrinks. well, we can continue to do what we do, create a new array/deque and copy all elements to it, but not having to copy would be much more efficent.

oprypin replied:
I would say that we indeed should never expose the `rewind` parameter and always do it.
Later someone can separately research whether that's a significant slowdown, and then possibly add this parameter. Also whether it's perhaps possible to do a 2-in-1 equivalent of `realloc` + `rewind`.

oprypin replied:
But, then again, potentially deprecating the `rewind` parameter to be ignored (always true) would also not be a breaking change. So maybe it's fine as is.

asterite replied:
Please let's not add the rewind parameter. Nothing in Array's API says anything about the internal implementation. This will prevent changing the internal implementation later on. Also users reading this will have no idea what this means.

Fryguy commented on src/array.cr:
```diff
@@ -2029,6 +2029,36 @@ class Array(T)
+  def ensure_capacity(capacity : Int32, rewind = true) : self
+    rewind() if rewind
+
+    if capacity - @offset_to_buffer >= @size
```
```suggestion
    if remaining_capacity >= @size
```

Fryguy commented on src/array.cr:
```diff
@@ -2029,6 +2029,36 @@ class Array(T)
+      resize_to_capacity capacity
+    else
+      err = "Insufficient capacity: #{capacity} cannot hold #{@size} elements"
+      err += " with an offset of #{@offset_to_buffer}" if @offset_to_buffer > 0
```
Similar to the discussion in https://github.com/crystal-lang/crystal/pull/11485/files#r756340774, this offset part of the error message is an internal implementation detail that users will likely not understand.

asterite commented:
Where did the requirement for introducing these methods come from? I don't doubt their usefulness, I just wonder if there are real programs where you tried this and it proved to be a benefit.

I'm only asking this because Ruby doesn't have such methods.

carlhoerberg commented:
We have Deques/Arrays that can grow to >100M elements, and then back to 0 or any value in between, and for our long running server application we want to minimize memory usage, hence what we do today when we notice that the size of the Deque/Array is significatly smaller than the capacity is to create a new Deque/Array and copies all the new remaining elements to it, so that the unused memory can be GCed, and returned to the OS.

asterite commented:
Wow! 100M elements in memory? Without knowing anything about what the application is... Can't that data be kept in a database? Something like sqlite3, for example.

straight-shoota commented:
@carlhoerberg So you occasionally want to trim down the capacity, but not as tight as the current size, in order to leave some room for growing again without resizing. Sounds legit.
Note however, that trimming does `realloc`, which may do a copy as well. So it's is not necessarily significantly faster than allocating a new array. It has an opportunity to be faster, depending on what `realloc` does.

beta-ziliani commented:
I removed the `rewind` parameters (it always rewind first), and I added an extra capacity to `trim_to_size` to have an option of leaving some spare room.

oprypin commented on src/array.cr:
```diff
@@ -2029,6 +2029,31 @@ class Array(T)
+    raise ArgumentError.new("Negative extra capacity: #{extra}") if extra < 0
+
+    rewind
+    resize_to_capacity (@size + extra)
```
Hmm there's an extra space here; surprised `crystal tool format` didn't complain

straight-shoota replied:
Interesting. I suppose that's because it's just interpreted as method call plus a separate parenthesized expression, not as a method call with args in parenthesis.

oprypin commented on src/array.cr:
```diff
@@ -2029,6 +2029,31 @@ class Array(T)
+    shift_buffer_by -@offset_to_buffer
+  end
+
+  # Enusures that the internal buffer has at least `capacity` elements.
```
Typo "ensures"

Sija commented on src/array.cr:
```diff
@@ -2029,6 +2029,31 @@ class Array(T)
+
+  # Enusures that the internal buffer has at least `capacity` elements.
+  def ensure_capacity(capacity : Int32) : self
+    rewind
```
This will do a realloc even if the given `capacity` is smaller than `@size`, is that intended?

straight-shoota replied:
It doesn't realloc, but memcpy. But that's the same problem. 
I'm not sure `#ensure_capacity` should ever rewind.

asterite replied:
There's never a need to rewind.

asterite commented on src/array.cr:
```diff
@@ -2029,6 +2029,31 @@ class Array(T)
+
+  # Reduces the internal buffer to exactly fit the number of elements in the
+  # array, plus `extra` elements.
+  def trim_to_size(extra : Int32 = 0) : self
```
This extra argument is pretty confusing. Trim to size 2 looks like it will trim it to size 2. I would rename extra to "plus", and make it a named argument. Then it will read like "trim to size plus 2".

oprypin replied:
Hmm I'm not sure I agree on the word. `extra` is fully correct English, I slightly prefer that one. "with 2 extra elements".
But yes, definitely would need to remove the ability to pass this as a positional argument :+1:

asterite replied:
Yes, it's mainly that as a positional argument a reader might be confused. The named argument could still be extra, but then it doesn't read like an English sentence. "trim to size extra 2" vs "trim to size plus 2". Maybe there's another word or set of words we could use.

beta-ziliani replied:
Vote:

 ❤️   `a.trim_to_size(plus: 2)` 
 😄 `a.trim_to_size(extra: 2)` 
 🎉  `a.trim_to_size(with_extra: 2)`

I can't think of another relevant names.

oprypin replied:
I dislike the idea of naming a parameter `with_anything`. My impression was that the "with" is implied by the opening of the list of arguments -- of course you're calling the method with something. I was going to make an argument that by that logic you could add `with_` to any parameter whatsoever, but actually I didn't find any examples, so it could just be that I'm wrong here.

Sija replied:
Maybe `bonus`? :)

Sija replied:
Or more technical `additive`.

HertzDevil commented:
Any reason that extra space can be reserved only on the right (for `#push`) but not on the left (for `#unshift`)?

beta-ziliani commented:
@HertzDevil yes: delay a proper treatment of `@offset_to_buffer`.

HertzDevil commented:
This implementation, as it stands right now, doesn't account for the buffer offset, so it breaks once elements are unshifted:

```crystal
x = Array.new(5, &.itself)
x           # => [0, 1, 2, 3, 4]
x.@capacity # => 5

x.unshift(5)
x           # => [5, 0, 1, 2, 3, 4]
x.@capacity # => 10

x.ensure_capacity(8)
x.push(6)
x           # => [5, 0, 1, 2, 0, 0, 6]
x.@capacity # => 8
```
